### DISCUSSION TITLE
Table with a column selector?

### DISCUSSION THREAD

**Discussion by amks1** [zauberzeug/nicegui#2654] [Q&A] [answered]
### Question

Hi, 
I'm considering recoding some of my frontend elements in NiceGUI. How can I get a table like this with a column selector? It can either be a `table` element or `aggrid`. In some cases I need to enable selection of a single column, and in others I need multiple columns.

![image](https://github.com/zauberzeug/nicegui/assets/68153785/65b3dd5f-c6a9-405b-9d9e-36cfec91aac1)

Any guidance will be much appreciated. 
Thanks!

**Comment by stdusr** [comment]
This?

```py
from nicegui import ui, app

grid = ui.aggrid({
    'defaultColDef': {'flex': 1},
    'columnDefs'   : [
        {'headerName': 'Name', 'field': 'name', "checkboxSelection": True, "headerCheckboxSelection": True},
        {'headerName': 'Age', 'field': 'age', "checkboxSelection": True, "headerCheckboxSelection": True},        
        ],
    'rowData'      : [
        {'name': 'Alice', 'age': 18, 'parent': 'David'},
        {'name': 'Bob', 'age': 21, 'parent': 'Eve'},
        {'name': 'Carol', 'age': 42, 'parent': 'Frank'},
        ],
    'rowSelection' : 'multiple',
    }).classes('max-h-40')

ui.run()


```

**Reply by amks1** [reply]
![image](https://github.com/zauberzeug/nicegui/assets/68153785/9259e931-77a6-4368-84cd-4a2db6741024)

Thanks, but this doesn't work for what I need: 
1. I'm unable to select a single column. Clicking on any header selects both columns. 
2. There are checkboxes in the rows as well, which I don't need.

I had been through the ag-grid documentation and there doesn't seem to be an easy way to select just columns. Is there another approach?

**Comment by falkoschindler** [comment]
Hi @amks1,

Here is a possible approach using `ui.table`:

```py
columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name', 'selected': True},
    {'name': 'age', 'label': 'Age', 'field': 'age', 'selected': False},
]
rows = [
    {'name': 'Alice', 'age': 18},
    {'name': 'Bob', 'age': 21},
    {'name': 'Carol'},
]

def handle_check(e: events.GenericEventArguments) -> None:
    for column in columns:
        if column['name'] == e.args['name']:
            column['selected'] = not column['selected']
    table.update()

table = ui.table(columns=columns, rows=rows, row_key='name')
table.add_slot('header-cell', r'''
    <q-th :props="props">
        <q-checkbox
            :label="props.col.label"
            :model-value="props.col.selected"
            @update:model-value="$parent.$emit('check', props.col)"
        />
    </q-th>
''')
table.on('check', handle_check)
```

It is rather advanced, using scoped slots and custom events. But I hope the code explains itself nonetheless.

**Reply by amks1** [reply]
Thanks so much, that works very well. For the table where I need to enable selection of a single column, I tried changing `q-checkbox` to `q-radio`. Although radio buttons are displayed on the frontend, they aren't clickable. :) What's the solution for that?

**Reply by falkoschindler** [reply, marked as answer]
I couldn't figure out how to do it with `q-radio`, because its internal state handling is a bit different to `q-checkbox`. But you can add
```
checked-icon="radio_button_checked"
unchecked-icon="radio_button_unchecked"
```
to the `q-checkbox` tag and change the event handler like this
```py
def handle_check(e: events.GenericEventArguments) -> None:
    for column in columns:
        if column['name'] == e.args['name'] and column['selected']:
            return
    for column in columns:
        column['selected'] = column['name'] == e.args['name']
    table.update()
```
to better reflect how radio buttons usually work.

**Reply by amks1** [reply]
Perfect, thanks a lot!!
